doc/manual/types-vs-contracts.md

```diff
@@ -0,0 +1,148 @@
+# Type vs contracts: when to?
+
+You are writing Nickel code and wonder how it should be annotated. Leave it
+alone? Add type annotations? Use contracts? Here is a quick guide when you don't
+know what to do!
+
+What is the nature of the expression you are considering?
+
+## A function
+
+Most of the time, you should use **type annotations** for functions. For
+reusable code, typing is often more adapted than contracts. You can
+exceptionally use contracts when types are not expressive enough to encode the
+property you want (such as in `#ValidUrl -> #Port -> #ValidUrl`) or if the type
+system is not powerful enough to see that your code is correct.
+
+Type annotations do have a runtime cost. If you hit contracts-related
+performances issues, you can always disable some contract checks using specific
+flags (TO BE PRECISED ONCE WE HAVE THOSE FLAGS). With annotations, code is still
+typechecked, and you can turn contracts checking back on for debugging. Without
+annotations, you're out of luck.
+
+What to do depends on the context:
+
+- *Anonymous function: nothing*. Short, anonymous functions can
+    usually live without annotation. Inside a typed block, they will be
+    typechecked anyway. Outside, anonymous function can't be reused elsewhere,
+    and are generally short functions passed as arguments to typed higher-order
+    functions, which will apply a guarding contract.
+
+    Example: `lists.map (fun x => x + 1) [1,2,3]`
+- *Let-bound function outside of typed block: use a type annotation.* Even if
+    local to a file, if your function is bound to a variable, it can be
+    potentially reused in different places.
+
+    Example: `let appendTm: Str -> Str = fun s => s ++ "(TM)" in ...`
+- *Let-bound function inside a typed block: nothing or type annotation*. Inside a
+    typed block, types are inferred, so it is OK for simple functions to not be
+    annotated. However, you are required to annotate it if it is polymorphic,
+    because the typechecker won't infer polymorphic types for you. When the
+    function type is non trivial, it can also be better to write an annotation
+    for the sake of clarity.
+
+    Example:
+    ```nickel
+    let foo : Num =
+      let addTwo = fun x => x + 2 in
+      addTwo 4
+    in ...
+
+    let foo : Num =
+      let ev : ((Num -> Num) -> Num) -> Num -> Num
+        = fun f x => f (functions.const x) in
+      ev (fun f => f 0) 1
+    in ...
+    ```
+
+## Library (record of functions)
```

Comment:
This is about a gotcha. This should go in a cookbook section somewhere. Maybe said cookbook should be in a different file indeed.

Reply:
Moved in a sink `cookbook.md` for now.